ocaml/xapi/repository_helpers.ml

```diff
@@ -1478,3 +1524,114 @@ let set_pending_guidances ~ops ~coming =
            to_be_removed ;
          do_with_vm_pending_guidances ~op:ops.vm_add ~vm:vm_ref_str to_be_added
      )
+
+let failure_of_livepatch_component = function
+  | Livepatch.Xen ->
+      Guidance.RebootHostOnXenLivePatchFailure
+  | Livepatch.Kernel ->
+      Guidance.RebootHostOnKernelLivePatchFailure
+
+let component_of_livepatch_failure = function
+  | Guidance.RebootHostOnXenLivePatchFailure ->
+      Some Livepatch.Xen
+  | Guidance.RebootHostOnKernelLivePatchFailure ->
+      Some Livepatch.Kernel
+  | _ ->
+      None
+
+let merge_livepatch_failures ~previous_failures ~applied ~failed =
+  let current_failures = List.map failure_of_livepatch_component failed in
+  (* Determine how to deal with previous livepatch failures *)
+  List.fold_left
+    (fun acc_fails prev_fail ->
+      match component_of_livepatch_failure prev_fail with
+      | Some prev_failed_component -> (
+          let comp_str = Livepatch.string_of_component prev_failed_component in
+          match
+            ( List.mem prev_fail current_failures
+            , List.mem prev_failed_component applied
+            )
+          with
+          | true, false ->
+              debug "%s Livepatch failed again." comp_str ;
+              acc_fails
+          | false, true ->
+              (* applied in this update. No changes on current failures *)
+              debug
+                "%s Livepatch (failed in previous updates) has been applied."
+                comp_str ;
+              acc_fails
+          | true, true ->
```

Comment:
When there are 2 xen livepatches, one applied successfully, one failed, will `Xen` exists both in `applied` and `failed`?

Reply:
It's impossible for two Xen live patches to be applied.